### PR TITLE
Add AuthenticationProperties.Parameters property bag

### PR DESCRIPTION
**Pull request by poke**
As discussed over in aspnet/Security#1691, this is adding a `Parameters` property bag to the `AuthenticationProperties` object which can be used to pass values to the authentication handler.

I’ve also made the helper methods for the `Items` collection protected to allow them to be reused in subclasses.

I also considered adding tests for the helpers since there were none; just say if you’d like that.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs [hunk 1]
```diff
@@ -20,27 +20,33 @@ public class AuthenticationProperties
...
             : this(items: null)
         {
         }
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="AuthenticationProperties"/> class
+        /// Initializes a new instance of the <see cref="AuthenticationProperties"/> class.
         /// </summary>
         /// <param name="items"></param>
         public AuthenticationProperties(IDictionary<string, string> items)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: We probably need an additional constructor that takes both.

**Comment by HaoK** [comment]
yes please to tests :)

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 1 above)
Comment (HaoK) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Do we want a copy constructor/overload that lets you easily copy both Items and Parameters into a new AuthenticationProperties?

Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Hmm, I didn’t want to add one for the `Properties` dictionary as that would likely be confusing but a copy constructor could be useful.

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Yeah either a copy constructor that takes AuthenticationProperties or one that takes items + properties, should have at least one of those options.

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Make it one that takes both collections, not a copy constructor. We may need this if we have to wrap Identitie's result.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs [hunk 2]
```diff
@@ -86,12 +92,12 @@ public string RedirectUri
             set => SetBool(RefreshKey, value);
         }
 
-        private string GetString(string key)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Make them public? They're operating on public state.

Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: All of them? Don’t want it to be too noisy and distract from the properties.

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Hmm, ok, leave it protected, but it still needs a basic doc comment.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs [hunk 3]
```diff
@@ -86,12 +92,12 @@ public string RedirectUri
...
+        protected string GetString(string key)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Two more methods to add: `public/protected object GetParameter(string name)` that does the same TryGetValue to avoid dictionary throwing. And add SetParemeter for symmetry.

Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Generic overload that attempts a type cast yes/no?

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: I think we should see what the actual derived ChallengeProperties look like before adding the sugar/helpers that we need for them...  But I don't disagree likely we'll need more helpers for the parameters here...

I just think we should start with none, and add them once we need them

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Basically I don't think we'll be able to properly review/design the base class helpers without seeing what the ChallengeProperties actually look like for real

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: We don't want to do too many trips back and forth, x-repo development is even more of a pain for external users. A single GetParameter method is better than nothing. Don't worry about the generic.

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: I'm not sure its any less painful for us :)  But I'd actually recommend doing both PRs in parallel instead of back and forth per say...

Start with the initial one to get the type to show up in the official bits (if needed), and then have a parallel one with additional helpers added as you go that gets checked in at the same time as security.

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: But given how this is purely additive, he can probably just create both PRs at the same time and we can merge them together at once they are fully complete

Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: We have at least an `ICollection<string>` for the scopes, a `TimeSpan?` for the OIDC’s `MaxAge`, and a `bool` for Google’s `IncludeGrantedScopesParameterKey`.

Having a generic `T GetParameter<T>(string key)` and a `void SetParameter<T>(string key, T value)` would easily allow us to cover all of this, so we wouldn’t need more helpers than those for the properties.

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Yeah that's how I'd have done it, sounds good to me, do what feels best/cleanest :)

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 1 above)
Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: The current constructor will not copy the passed `Items` though but will use the passed collection as is. Would you want the same behavior for the `Parameters` dictionary then? This would be kind of the opposite of a copy constructor then.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 3 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: The generic only hides a cast. It's not giving you serialization like the other helpers here.

`GetParameter<ICollection<string>>` vs `(ICollection<string>)GetParameter` is pretty much the same.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 1 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Yes, take the existing collection. It's a very shallow copy 😁

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 3 above)
Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Going back to my point...just implement the actual properties classes...before trying to design the helpers :)

Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: I was actually doing the following in my implementation:

```
Parameters.TryGetValue(key, out var obj) && obj is T value ? value : default
```

I would personally rather ignore parameters that have the wrong type than have them raise an exception for them (or to add error handling into the handlers for everything).

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: My personal guess is that we would have a common set of helpers, that takes the Dictionary as an argument.

`GetString(Items | Parameters, key),`

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: The dictionaries are different types...

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs [hunk 4]
```diff
@@ -109,6 +127,40 @@ protected void SetString(string key, string value)
...
+        /// <typeparam name="T">Parameter type,</typeparam>
+        /// <param name="key">Parameter key.</param>
+        /// <param name="value">Value to set or <c>null</c> to remove the property.</param>
+        public void SetParameter<T>(string key, T value)
+        {
+            if (value != null)
+            {
+                Parameters[key] = value;
+            }
+            else if (Parameters.ContainsKey(key))
```
Comment (HaoK) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Do we really need this null clearing logic?  Can't they just do Parameters.Remove(key) directly instead of set with null?

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs [hunk 5]
```diff
@@ -109,6 +127,40 @@ protected void SetString(string key, string value)
...
 
+        /// <summary>
+        /// Get a parameter from the <see cref="Parameters"/> collection.
+        /// </summary>
+        /// <typeparam name="T">Parameter type.</typeparam>
+        /// <param name="key">Parameter key.</param>
+        /// <returns>Retrieved value or the default value if the property is not set.</returns>
+        public T GetParameter<T>(string key)
+        {
+            return Parameters.TryGetValue(key, out var obj) && obj is T value ? value : default;
```
Comment (HaoK) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: nit if you are going to right a fancy complex one liner, you might as well use => and drop the return and get rid of the extra { }'s :)

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 4 above)
Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: That was mostly to get a symmetric behavior to the Set~ methods on the `Items` dictionary.

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: I don't think that's an important concern as the Items are used for serialization where this behavior is much more important

Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: That’s a good point! I’ll remove that part then.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs [hunk 6]
```diff
@@ -20,27 +20,41 @@ public class AuthenticationProperties
...
+            Parameters = parameters ?? new Dictionary<string, object>(StringComparer.Ordinal);
         }
 
         /// <summary>
         /// State values about the authentication session.
         /// </summary>
         public IDictionary<string, string> Items { get; }
 
+        /// <summary>
+        /// Collection of parameters that are passed to the authentication handler.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: "These are not intended for serialization or persistence, only for flowing data between call sites."

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 2 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Why are we exposing these things?

**Comment by poke** [comment]
Alright, squashed and cleaned up! I guess this is ready if @HaoK has no objections.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 2 above)
Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: (GitHub’s UI sometimes makes it really easy to miss comments…)

@davidfowl I made the string-based helpers public so that they can serve as a more friendly way to interact with the `Items` dictionary. And especially so that handlers have a simple API to retrieve values (for when they actually want to persist data inside `Items`).

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: I thought we were going to keep these as protected?

Comment (HaoK) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: We don't actually need Get/SetParameter public either right?  We are going to have properties that expose the parameters right?

**Comment by Tratcher** [comment]
I tried moving your PR to the release/2.1 branch but it needs a clean rebase. We'll be working in release/2.1 for the remainder of preview2.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 2 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: We don't do this anywhere else in the stack.

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: We do, but they're usually extension methods.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: Where?

Comment (poke) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: I don’t mind either way but as soon as a handler needs a value from the items collection, they are effectively going to reimplement the thing again. And since the `GetParameter` method is public, I don’t think a set of public methods for the `Items` collection hurts.

**Comment by poke** [comment]
Rebased onto release/2.1.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs (hunk 2 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs]: https://github.com/aspnet/Caching/blob/dev/src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs

**Comment by Tratcher** [comment]
@poke can you send me your e-mail address to the address in my profile? I have something to send you.

**Comment by poke** [comment]
@Tratcher Btw. did you receive my email? I never got anything back (I’m actually losing trust in gmail at the moment >_<)

**Comment by Tratcher** [comment]
Ack, it's taking longer to put together than I expected.

**Comment by poke** [comment]
Alright, just wanted to make sure that it wasn’t my email not working properly :)
